src/main/java/com/icthh/xm/gate/security/oauth2/IdpAuthenticationSuccessHandler.java

```diff
@@ -0,0 +1,176 @@
+import static com.icthh.xm.gate.config.Constants.HEADER_TENANT;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import com.icthh.xm.commons.domain.idp.model.IdpPublicConfig.IdpConfigContainer.Features;
+import com.icthh.xm.commons.tenant.TenantContextHolder;
+import io.github.jhipster.config.JHipsterProperties;
+
+import java.io.IOException;
+import java.util.Base64;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import lombok.extern.slf4j.Slf4j;
+
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.core.ParameterizedTypeReference;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.oauth2.core.oidc.OidcIdToken;
+import org.springframework.security.oauth2.core.oidc.user.OidcUser;
+import org.springframework.security.web.authentication.AuthenticationSuccessHandler;
+import org.springframework.stereotype.Component;
+import org.springframework.util.LinkedMultiValueMap;
+import org.springframework.util.MultiValueMap;
+import org.springframework.web.client.RestTemplate;
+
+/**
+ * XM Strategy used to handle a successful IDP user authentication.
+ */
+@Slf4j
+@Component
+public class IdpAuthenticationSuccessHandler implements AuthenticationSuccessHandler {
+
+    public static final String GRANT_TYPE_ATTR = "grant_type";
+    public static final String GRANT_TYPE_IDP_TOKEN = "idp_token";
+    public static final String TOKEN_ATTR = "token";
+    public static final String COLON_SEPARATOR = ":";
+
+    private final ObjectMapper objectMapper;
+    private final RestTemplate restTemplate;
+    private final TenantContextHolder tenantContextHolder;
+    private final IdpConfigRepository idpConfigRepository;
+    private final JHipsterProperties jhipsterProperties;
+
+    public IdpAuthenticationSuccessHandler(ObjectMapper objectMapper,
+                                           @Lazy @Qualifier("loadBalancedRestTemplate") RestTemplate restTemplate,
+                                           TenantContextHolder tenantContextHolder,
+                                           IdpConfigRepository idpConfigRepository,
+                                           JHipsterProperties jhipsterProperties) {
+        this.objectMapper = objectMapper;
+        this.tenantContextHolder = tenantContextHolder;
+        this.restTemplate = restTemplate;
+        this.idpConfigRepository = idpConfigRepository;
+        this.jhipsterProperties = jhipsterProperties;
+    }
+
+    @Override
+    public void onAuthenticationSuccess(HttpServletRequest request,
```

Comment:
can we cover this method by lep? (if it possible)

Reply:
@sergeysenja1992 we decided not to add LEP here from a start as we do not see the cases of using it.
But in the future, we may get back to this question and add if need.